IllegalArgumentException for an empty `application/json` body

If you use `muuntaja.middleware/wrap-format-request` and do a HTTP request with the header 'Content-Type: application/json` and with an empty body, Muuntaja throws an IllegalArgumentException. I think that empty bodies should be just ignored.

A stacktrace from a Kekkonen hello world app:

```
clojure.lang.ExceptionInfo: Malformed application/json request. {:type :muuntaja/decode, :default-format "application/json", :format "application/json", :charset "utf-8", :request {:remote-addr "0:0:0:0:0:0:0:1", :params {}, :headers {"origin" "http://localhost:5000", "host" "localhost:5000", "content-type" "application/json", "content-length" "0", "referer" "http://localhost:5000/api-docs/index.html", "connection" "keep-alive", "accept" "application/json", "accept-language" "fi-fi", "accept-encoding" "gzip, deflate", "dnt" "1", "kekkonen.mode" "invoke"}, :async-channel #object[org.httpkit.server.AsyncChannel 0x37ba91df "/0:0:0:0:0:0:0:1:5000<->/0:0:0:0:0:0:0:1:49756"], :server-port 5000, :muuntaja/request #FormatAndCharset{:format "application/json", :charset "utf-8"}, :content-length 0, :form-params {}, :websocket? false, :query-params {}, :content-type "application/json", :character-encoding "utf8", :uri "/api/sample/inc!", :server-name "localhost", :query-string nil, :muuntaja/response #FormatAndCharset{:format "application/json", :charset "utf-8"}, :body nil, :scheme :http, :request-method :post}}
	at clojure.core$ex_info.invokeStatic(core.clj:4617)
	at clojure.core$ex_info.invoke(core.clj:4617)
	at muuntaja.core$fail_on_request_decode_exception.invokeStatic(core.clj:236)
	at muuntaja.core$fail_on_request_decode_exception.invoke(core.clj:230)
	at muuntaja.core$decode_request_body.invokeStatic(core.clj:369)
	at muuntaja.core$decode_request_body.invoke(core.clj:363)
	at muuntaja.core$decode_request.invokeStatic(core.clj:379)
	at muuntaja.core$decode_request.invoke(core.clj:376)
	at muuntaja.core$format_request.invokeStatic(core.clj:387)
	at muuntaja.core$format_request.invoke(core.clj:384)
	at muuntaja.middleware$wrap_format_request$fn__31336.invoke(middleware.clj:113)
	at kekkonen.middleware$wrap_exceptions$fn__32593.invoke(middleware.clj:74)
	at muuntaja.middleware$wrap_format_response$fn__31340.invoke(middleware.clj:131)
	at muuntaja.middleware$wrap_format_negotiate$fn__31333.invoke(middleware.clj:95)
	at kekkonen.middleware$wrap_keyword_keys$fn__32602.invoke(middleware.clj:106)
	at ring.middleware.keyword_params$wrap_keyword_params$fn__30330.invoke(keyword_params.clj:35)
	at ring.middleware.nested_params$wrap_nested_params$fn__30502.invoke(nested_params.clj:86)
	at ring.middleware.params$wrap_params$fn__30592.invoke(params.clj:64)
	at clojure.lang.Var.invoke(Var.java:379)
	at org.httpkit.server.HttpHandler.run(RingHandler.java:91)
	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
	at java.lang.Thread.run(Thread.java:745)
Caused by: clojure.lang.ExceptionInfo: Malformed application/json in :muuntaja/decode {:type :muuntaja/decode, :format "application/json"}
	at clojure.core$ex_info.invokeStatic(core.clj:4617)
	at clojure.core$ex_info.invoke(core.clj:4617)
	at muuntaja.core$on_exception.invokeStatic(core.clj:66)
	at muuntaja.core$on_exception.invoke(core.clj:64)
	at muuntaja.core$create_coder$f__31187.invoke(core.clj:106)
	at muuntaja.core$decode_request_body.invokeStatic(core.clj:367)
	... 20 more
Caused by: java.lang.IllegalArgumentException: No implementation of method: :as-input-stream of protocol: #'muuntaja.protocols/AsInputStream found for class: nil
	at clojure.core$_cache_protocol_fn.invokeStatic(core_deftype.clj:568)
	at clojure.core$_cache_protocol_fn.invoke(core_deftype.clj:560)
	at muuntaja.protocols$eval30740$fn__30741$G__30731__30746.invoke(protocols.clj:45)
	at muuntaja.core$create_coder$f__31187.invoke(core.clj:104)
	... 21 more
```

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -1,9 +1,9 @@
 ## UNRELEASED
 
 * **BREAKING**: Handling empty responses
-  * `:allow-empty-input-on-decode?` is now called `:allow-empty-input`, defaulting to `true`
-  * by default, empty input(streams) is decoded into`nil`
-  * setting the value to `false` with cause the decoder to do whatever it does
+  * `:allow-empty-input-on-decode?` is now called `:allow-empty-input?`. It's a boolean:
+    * `true` (default): empty input(stream) is decoded into `nil`
+    * `false` with cause the decoder to do whatever it does (e.g. Transit fails, Cheshire returns `nil`)
 
 * updated deps:
 

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -1,3 +1,12 @@
+## UNRELEASED
+
+* `nil` input is always resolved as `nil`. Fixes [#41](https://github.com/metosin/muuntaja/issues/41)
+* updated deps:
+
+```clj
+[cheshire "5.7.1"] is available but we use "5.7.0"
+```
+
 ## 0.2.1 (2.4.2017)
 
 * removed direct dependencies to msgpack, fixes [#39](https://github.com/metosin/muuntaja/issues/39).

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -1,6 +1,10 @@
 ## UNRELEASED
 
-* `nil` input is always resolved as `nil`. Fixes [#41](https://github.com/metosin/muuntaja/issues/41)
+* **BREAKING**: Handling empty responses
+  * `:allow-empty-input-on-decode?` is now called `:allow-empty-input`, defaulting to `true`
+  * by default, empty input(streams) is decoded into`nil`
+  * setting the value to `false` with cause the decoder to do whatever it does
+
 * updated deps:
 
 ```clj

--- a/README.md
+++ b/README.md
@@ -184,7 +184,7 @@ be used in the response pipeline.
         :decode-request-body? (constantly true)
         :encode-reseponse-body? encode-collections-with-override}
 
- :allow-empty-input-on-decode? false
+ :allow-empty-input? true
 
  :default-charset "utf-8"
  :charsets muuntaja/available-charsets

--- a/project.clj
+++ b/project.clj
@@ -8,7 +8,7 @@
   :source-paths ["src/clj"]
   :javac-options ["-Xlint:unchecked"]
   :java-source-paths ["src/java"]
-  :dependencies [[cheshire "5.7.0" :exclusions [com.fasterxml.jackson.core/jackson-core]]
+  :dependencies [[cheshire "5.7.1" :exclusions [com.fasterxml.jackson.core/jackson-core]]
                  [com.fasterxml.jackson.core/jackson-databind "2.8.7"]
                  [com.cognitect/transit-clj "0.8.300"]]
   :plugins [[lein-codox "0.10.3"]]
@@ -17,7 +17,7 @@
           :defaults {:doc/format :markdown}}
   :profiles {:dev {:jvm-opts ^:replace ["-server"]
                    :dependencies [[org.clojure/clojure "1.8.0"]
-                                  [ring/ring-core "1.6.0-RC1"]
+                                  [ring/ring-core "1.6.0"]
                                   [ring-middleware-format "0.7.2"]
                                   [ring-transit "0.1.6"]
                                   [ring/ring-json "0.4.0"]

--- a/src/clj/muuntaja/core.clj
+++ b/src/clj/muuntaja/core.clj
@@ -132,6 +132,9 @@
           (str "Invalid default format " default-format)
           {:formats valid-format?
            :default-format default-format})))
+    (assert
+      (not (:allow-empty-input-on-decode? options))
+      ":allow-empty-input-on-decode? is deprecated in Muuntaja 0.3.0")
     (->
       (merge
         (dissoc options :formats)

--- a/src/clj/muuntaja/core.clj
+++ b/src/clj/muuntaja/core.clj
@@ -83,8 +83,7 @@
                 (if (and allow-empty-input? decode?)
                   (fn [^InputStream is charset]
                     (if (pos? (.available is)) ($ is charset)))
-                  (fn [^InputStream is charset]
-                    ($ is charset))))
+                  $))
         prepare (if decode? protocols/as-input-stream identity)]
     (if (and p pf)
       (fn f

--- a/src/clj/muuntaja/core.clj
+++ b/src/clj/muuntaja/core.clj
@@ -69,7 +69,7 @@
        :format format}
       e)))
 
-(defn- create-coder [format type spec spec-opts default-charset allow-empty-input-on-decode? [p pf]]
+(defn- create-coder [format type spec spec-opts default-charset allow-empty-input? [p pf]]
   (let [decode? (= type :muuntaja/decode)
         g (as-> spec $
 
@@ -80,11 +80,11 @@
                   $)
 
                 ;; optional guard on empty imput
-                (if (and allow-empty-input-on-decode? decode?)
+                (if (and allow-empty-input? decode?)
                   (fn [^InputStream is charset]
-                    (if (and (not (nil? is)) (pos? (.available is))) ($ is charset)))
+                    (if (pos? (.available is)) ($ is charset)))
                   (fn [^InputStream is charset]
-                    (if-not (nil? is) ($ is charset)))))
+                    ($ is charset))))
         prepare (if decode? protocols/as-input-stream identity)]
     (if (and p pf)
       (fn f
@@ -106,7 +106,7 @@
            (catch Exception e
              (on-exception e format type))))))))
 
-(defn- create-adapters [formats default-charset allow-empty-input-on-decode?]
+(defn- create-adapters [formats default-charset allow-empty-input?]
   (->> (for [[format {:keys [decoder decoder-opts encoder encoder-opts encode-protocol]}] formats]
          (if-not (or encoder decoder)
            (throw
@@ -116,16 +116,16 @@
                 :formats (keys formats)}))
            [format (records/map->Adapter
                      (merge
-                       (if decoder {:decode (create-coder format :muuntaja/decode decoder decoder-opts default-charset allow-empty-input-on-decode? nil)})
-                       (if encoder {:encode (create-coder format :muuntaja/encode encoder encoder-opts default-charset allow-empty-input-on-decode? encode-protocol)})))]))
+                       (if decoder {:decode (create-coder format :muuntaja/decode decoder decoder-opts default-charset allow-empty-input? nil)})
+                       (if encoder {:encode (create-coder format :muuntaja/encode encoder encoder-opts default-charset allow-empty-input? encode-protocol)})))]))
        (into {})))
 
 (declare default-options)
 (declare http-create)
 
 (defn- -create
-  [{:keys [formats default-format default-charset allow-empty-input-on-decode?] :as options}]
-  (let [adapters (create-adapters formats default-charset allow-empty-input-on-decode?)
+  [{:keys [formats default-format default-charset allow-empty-input?] :as options}]
+  (let [adapters (create-adapters formats default-charset allow-empty-input?)
         valid-format? (key-set formats identity)]
     (when-not (or (not default-format) (valid-format? default-format))
       (throw
@@ -211,7 +211,7 @@
           :decode-request-body? (constantly true)
           :encode-response-body? encode-collections-with-override}
 
-   :allow-empty-input-on-decode? false
+   :allow-empty-input? true
 
    :default-charset "utf-8"
    :charsets available-charsets

--- a/src/clj/muuntaja/core.clj
+++ b/src/clj/muuntaja/core.clj
@@ -82,8 +82,9 @@
                 ;; optional guard on empty imput
                 (if (and allow-empty-input-on-decode? decode?)
                   (fn [^InputStream is charset]
-                    (if (pos? (.available is)) ($ is charset)))
-                  $))
+                    (if (and (not (nil? is)) (pos? (.available is))) ($ is charset)))
+                  (fn [^InputStream is charset]
+                    (if-not (nil? is) ($ is charset)))))
         prepare (if decode? protocols/as-input-stream identity)]
     (if (and p pf)
       (fn f

--- a/src/clj/muuntaja/protocols.clj
+++ b/src/clj/muuntaja/protocols.clj
@@ -55,4 +55,8 @@
 
   String
   (as-input-stream [this]
-    (ByteArrayInputStream. (.getBytes this "utf-8"))))
+    (ByteArrayInputStream. (.getBytes this "utf-8")))
+
+  nil
+  (as-input-stream [_]
+    nil))

--- a/src/clj/muuntaja/protocols.clj
+++ b/src/clj/muuntaja/protocols.clj
@@ -59,4 +59,4 @@
 
   nil
   (as-input-stream [_]
-    nil))
+    (ByteArrayInputStream. (byte-array 0))))

--- a/test/muuntaja/core_test.clj
+++ b/test/muuntaja/core_test.clj
@@ -225,3 +225,8 @@
                 (assoc-in
                   [:formats "application/jsonz" :decoder-opts]
                   {:keywords? false})))))))
+
+(deftest migration-to-0.3.0-test
+  (is (thrown?
+        AssertionError
+        (m/create (assoc m/default-options :allow-empty-input-on-decode? true)))))

--- a/test/muuntaja/core_test.clj
+++ b/test/muuntaja/core_test.clj
@@ -71,11 +71,15 @@
                    (yaml-format/with-yaml-format)
                    (assoc :allow-empty-input-on-decode? true)))]
 
-      (testing "by default - exception is thrown"
+      (testing "by default - exception is thrown for empty stream"
         (is (thrown? Exception (m/decode m "application/transit+json" (empty)))))
 
-      (testing "optionally nil is returned"
-        (is (nil? (m/decode m2 "application/transit+json" (empty)))))
+      (testing "by default - nil input returns nil stream"
+        (is (nil? (m/decode m "application/transit+json" nil))))
+
+      (testing "optionally nil is returned in both cases"
+        (is (nil? (m/decode m2 "application/transit+json" (empty))))
+        (is (nil? (m/decode m2 "application/transit+json" nil))))
 
       (testing "all formats"
         (testing "with defaults"

--- a/test/muuntaja/core_test.clj
+++ b/test/muuntaja/core_test.clj
@@ -69,39 +69,39 @@
                (-> m/default-options
                    (msgpack-format/with-msgpack-format)
                    (yaml-format/with-yaml-format)
-                   (assoc :allow-empty-input-on-decode? true)))]
+                   (assoc :allow-empty-input? false)))]
 
-      (testing "by default - exception is thrown for empty stream"
-        (is (thrown? Exception (m/decode m "application/transit+json" (empty)))))
+      (testing "by default - nil is returned for empty stream"
+        (is (nil? (m/decode m "application/transit+json" (empty)))))
 
       (testing "by default - nil input returns nil stream"
         (is (nil? (m/decode m "application/transit+json" nil))))
 
-      (testing "optionally nil is returned in both cases"
-        (is (nil? (m/decode m2 "application/transit+json" (empty))))
-        (is (nil? (m/decode m2 "application/transit+json" nil))))
+      (testing "optionally decoder can decide to throw"
+        (is (thrown? Exception (m/decode m2 "application/transit+json" (empty))))
+        (is (thrown? Exception (m/decode m2 "application/transit+json" nil))))
 
       (testing "all formats"
-        (testing "with defaults"
+        (testing "with :allow-empty-input? false"
 
           (testing "json & yaml return nil"
             (are [format]
-              (= nil (m/decode m format (empty)))
+              (= nil (m/decode m2 format (empty)))
               "application/json"
               "application/x-yaml"))
 
           (testing "others fail"
             (are [format]
-              (thrown-with-msg? Exception #"Malformed" (m/decode m format (empty)))
+              (thrown-with-msg? Exception #"Malformed" (m/decode m2 format (empty)))
               "application/edn"
               "application/msgpack"
               "application/transit+json"
               "application/transit+msgpack")))
 
-        (testing "with :allow-empty-input-on-decode? true"
+        (testing "with defaults"
           (testing "all formats return nil"
             (are [format]
-              (= nil (m/decode m2 format (empty)))
+              (= nil (m/decode m format (empty)))
               "application/json"
               "application/edn"
               "application/x-yaml"

